lib/OpenQA/WebAPI/Controller/Admin/Needle.pm

```diff
@@ -77,18 +77,46 @@ sub ajax {
             my ($needles) = @_;
             my @data;
             my %modules;
+            my %needle_dir_real_paths;
 
-            my %needle_dir_map = map { $_->path => $_ } $needles->all;
-
+            my $schema = $self->schema;
             while (my $n = $needles->next) {
-                my $hash = {
+                my $filename = $n->filename;
+                my $hash     = {
                     id        => $n->id,
                     directory => $n->directory->name,
-                    filename  => $n->filename,
+                    filename  => $filename,
                 };
+                my $needle_dir_path = $n->directory->path;
+                my $real_needle_dir_id;
+
+                if ($needle_dir_real_paths{$needle_dir_path}) {
+                    my $real_path = $needle_dir_real_paths{$needle_dir_path}->{real_path};
+                    $real_needle_dir_id = $needle_dir_real_paths{$needle_dir_path}->{real_path_id}
+                      if $real_path && $needle_dir_path ne $real_path;
+                }
+                else {
+                    my $real_path_id         = $n->directory->id;
+                    my $needle_dir_real_path = realpath($needle_dir_path);
+                    if ($needle_dir_real_path && $needle_dir_real_path ne $needle_dir_path) {
+                        my $real_needle_dir
+                          = $self->schema->resultset('NeedleDirs')->find({path => $needle_dir_real_path});
```

Comment:
Better save `$self->schema->resultset('NeedleDirs')` and `$schema->resultset('Needles')` (which you're using later on) in a variable, e.g. when `$schema` is defined. That improves the line-wrapping *and* saves up to 4 method calls in each loop iteration.